OpenRA.Mods.Common/Widgets/Logic/Lobby/MapPreviewLogic.cs

```diff
@@ -44,220 +51,256 @@ public class MapPreviewLogic : ChromeLogic
+				preview.ShowUnoccupiedSpawnpoints = showUnoccupiedSpawnpoints;
 
-			var invalid = widget.GetOrNull("MAP_INVALID");
-			if (invalid != null)
-			{
-				invalid.IsVisible = () =>
+				var titleLabel = parent.GetOrNull<LabelWithTooltipWidget>("MAP_TITLE");
+				if (titleLabel != null)
 				{
-					var (map, serverStatus) = getMap();
-					return map.Status == MapStatus.Available && (serverStatus & Session.MapStatus.Incompatible) != 0;
-				};
+					titleLabel.IsVisible = () => getMap().Map != MapCache.UnknownMap;
+					var titleCache = new CachedTransform<string, string>(str =>
+					{
+						var truncatedText = WidgetUtils.TruncateText(str, titleLabel.Bounds.Width, Game.Renderer.Fonts[titleLabel.Font]);
 
-				SetupWidgets(invalid, modData, getMap, onMouseDown, getSpawnOccupants, getDisabledSpawnPoints, showUnoccupiedSpawnpoints);
-			}
+						if (str != truncatedText)
+							titleLabel.GetTooltipText = () => str;
+						else
+							titleLabel.GetTooltipText = null;
 
-			var validating = widget.GetOrNull("MAP_VALIDATING");
-			if (validating != null)
-			{
-				validating.IsVisible = () =>
-				{
-					var (map, serverStatus) = getMap();
-					return map.Status == MapStatus.Available && (serverStatus & Session.MapStatus.Validating) != 0;
-				};
+						return truncatedText;
+					});
+
+					titleLabel.GetText = () => titleCache.Update(getMap().Map.Title);
+				}
 
-				SetupWidgets(validating, modData, getMap, onMouseDown, getSpawnOccupants, getDisabledSpawnPoints, showUnoccupiedSpawnpoints);
+				return parent;
 			}
 
-			var download = widget.GetOrNull("MAP_DOWNLOADABLE");
-			if (download != null)
+			var previewLarge = SetupPreview("LARGE");
+			var previewSmall = SetupPreview("SMALL");
+
+			// Setup other preview widgets
+			Widget SetupAuthorAndMapType(string container)
 			{
-				download.IsVisible = () => getMap().Map.Status == MapStatus.DownloadAvailable;
+				var parent = widget.Get(container);
 
-				SetupWidgets(download, modData, getMap, onMouseDown, getSpawnOccupants, getDisabledSpawnPoints, showUnoccupiedSpawnpoints);
+				var typeLabel = parent.Get<LabelWidget>("MAP_TYPE");
+				var type = new CachedTransform<MapPreview, string>(
+					m => m.Categories.FirstOrDefault() ?? "");
 
-				var install = download.GetOrNull<ButtonWidget>("MAP_INSTALL");
-				if (install != null)
-				{
-					install.OnClick = () =>
-					{
-						getMap().Map.Install(mapRepository, () =>
-						{
-							if (orderManager != null)
-								Game.RunAfterTick(() => orderManager.IssueOrder(Order.Command($"state {Session.ClientState.NotReady}")));
-						});
-					};
-
-					install.IsHighlighted = () => installHighlighted;
-				}
+				typeLabel.GetText = () => type.Update(getMap().Map);
+
+				var authorLabel = parent.Get<LabelWidget>("MAP_AUTHOR");
+				var truncateCache = new CachedTransform<MapPreview, string>(
+					m => WidgetUtils.TruncateText(authorCache.Update(m.Author), authorLabel.Bounds.Width, Game.Renderer.Fonts[authorLabel.Font]));
```

Comment:
`Game.Renderer.Fonts[authorLabel.Font]` can be cached outside the CachedTransform

Reply:
If you insist